### ISSUE TITLE
feature requests from kernel programmers

### ISSUE THREAD

**Issue by andres-erbsen** [mit-plv/bedrock2#17]
- better syntax for struct access
- type-based pointer arithmetic
- easy declaration of external functions, with support for including relevant header files during printing of C code
- story for statically allocated (/external) variables / linker symbols
- debugger plugin for printing bedrock structs?

not really big issue:
- lack of break, continue, return (as long as require exists)
- purely syntactic differences

**Comment by samuelgruetter** [comment]
Which of these can be implemented in Gallina on top of the existing semantics, and which require changes to the existing semantics?

**Comment by andres-erbsen** [comment]
I guess that's a question for use to decide. I think it would be reasonable to first attempt "typerock" as a language that desugars to bedrock, but has actual types for pointers and structs. I currently don't see a nice way to do (1) and (2) with just Gallina functions returning bedrock code.

The rest are just tooling questions afais.

**Comment by gmalecha** [comment]
I think that `break`, `continue`, and `return` should be implemented as part of the language. While they can be implemented via desugaring, the desugaring introduces variables which complicate verification conditions. Enriching the weakest pre-condition calculus for these constructs is fairly simple, but the denotational semantics might be slightly more complex due to the current style of the definition. As far as `return`, I (personally) prefer the syntax `return : list expr -> cmd` rather than simply `return : cmd` (where variables are returned through names the way they are now). This is more in the style of systems programming languages such as rust, go, and C/C++.

The linker symbols and externals needs to be added as a notion of a compilation unit (which includes both functions and data). This can mostly be written on top of the existing compiler, though I think it would be more uniform to add a syntactic construct for "global_reference" to the expression language and handle everything through that. I wrote a sketch on top of an alternate semantics and it wasn't too difficult. The interesting part comes from defining linking.

**Comment by gmalecha** [comment]
Another feature request was stack-allocated structs. I assume that this won't be supported, but just a note.

**Comment by samuelgruetter** [comment]
Do you mean *addressable* stack-allocated structs?

**Comment by gmalecha** [comment]
That would be ideal, otherwise you have to pass them by value to actually use them which isn't very useful.

As I said, I assume you don't plan to support this.

**Comment by samuelgruetter** [comment]
Yeah, I don't have plans to support addressable stack-allocated structs. But I just asked because if you think non-addressable stack-allocated structs would be useful, I think they could be implemented in Gallina as syntactic sugar.
